File: 872/leaf-similar-trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
function TreeNode(val) {
    this.val = val;
    this.left = this.right = null;
}

/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function (root1, root2) {
    let arr1 = [];
    let arr2 = [];
    getLeaves(root1, arr1);
    getLeaves(root2, arr2);
    if (arr1.length != arr2.length)
        return false;
    else {
        while (arr1.length) {
            let item1 = arr1.pop();
            let item2 = arr2.pop();
            if (item1 != item2)
                return false;
        }
        return true;
    }
}

function getLeaves(root, leaves) {
    if (!root) return;
    if (!root.left && !root.right) {
        leaves.push(root.val);
    }
    getLeaves(root.left, leaves);
    getLeaves(root.right, leaves);
}

let root1 = new TreeNode(1);
let root2 = new TreeNode(2);
console.log(leafSimilar(root1, root2));

File: 872/LeafSimilarTrees.scala
object LeafSimilarTrees {
  class TreeNode(val _value: Int = 0) {
    var value: Int = _value
    var left: TreeNode = null
    var right: TreeNode = null
  }

  def leafSimilar(root1: TreeNode, root2: TreeNode): Boolean = {
    def getLeaves(root: TreeNode): List[Int] = {
      if (root == null) List.empty[Int]
      else if (root.left == null && root.right == null) List(root.value)
      else getLeaves(root.left) ++ getLeaves(root.right)
    }
    getLeaves(root1) == getLeaves(root2)
  }

  def main(args: Array[String]): Unit = {
    val root1 = new TreeNode(1)
    val root2 = new TreeNode(2)
    println(leafSimilar(root1, root2))
  }

}
